shared/markdown/parser.pegjs

```diff
@@ -1,154 +1,161 @@
+// TODO: should coalesce multiple line quotes
 QuoteBlock
- = QuoteBlockMarker _? children:FromQuote* LineTerminatorSequence { return {type: 'quote-block', children}; }
+ = QuoteBlockMarker WhiteSpace* children:__INLINE_MACRO__<!LineTerminatorSequence> LineTerminatorSequence { return {type: 'quote-block', children} }
 
 Bold
- = BoldMarker children:FromBold* BoldMarker { return {type: 'bold', children}; }
+ = BoldMarker !WhiteSpace children:__INLINE_MACRO__<!BoldMarker> BoldMarker !(BoldMarker / NormalChar) { return {type: 'bold', children: flatten(children)} }
 
 Italic
- = ItalicMarker children:FromItalic* ItalicMarker { return {type: 'italic', children}; }
+ = ItalicMarker !WhiteSpace children:__INLINE_MACRO__<!ItalicMarker> ItalicMarker !(ItalicMarker / NormalChar) { return {type: 'italic', children: flatten(children)} }
 
 Strike
- = StrikeMarker children:FromStrike* StrikeMarker { return {type: 'strike', children}; }
+ = StrikeMarker !WhiteSpace children:__INLINE_MACRO__<!StrikeMarker> StrikeMarker !(StrikeMarker / NormalChar) { return {type: 'strike', children: flatten(children)} }
 
 CodeBlock
- = Ticks3 children:InsideCodeBlock* Ticks3 { return {type: 'code-block', children}; }
+ = Ticks3 children:(!Ticks3 .)+ Ticks3 { return {type: 'code-block', children: flatten(children)} }
 
 InlineCode
- = Ticks1 children:InsideInlineCode* Ticks1 { return {type: 'inline-code', children}; }
+ = Ticks1 children:(!Ticks1 .)+ Ticks1 { return {type: 'inline-code', children: flatten(children)} }
 
-Emoji
- = EmojiMarker children:InsideEmojiMarker+ tone:InsideEmojiTone? ":" { return {type: 'emoji', children: [children.join('') + (tone || '')]}; }
+// Here we use the literal ":" because we want to not match the :foo in ::foo
+InsideEmojiMarker
+ = !EmojiMarker [a-zA-Z0-9+_-] { return text() }
 
-Text "text"
- = _? NonBlank+ _? {
-    return convertLink(text())
- }
+InsideEmojiTone
+ = "::skin-tone-" [1-6] { return text() }
 
-// Useful helpers
+Emoji
+ = EmojiMarker children:InsideEmojiMarker+ tone:InsideEmojiTone? EmojiMarker { return {type: 'emoji', children: [text()]} }
+
+NativeEmojiCharacter "unicode emoji"
+ = [__EMOJI_CHARACTERS__]
+
+NativeEmoji
+ = emoji:(NativeEmojiCharacter+)
+ {
+   const emojiText = emoji.join('')
+   const results = []
+   let match
+   let idx = 0
+   while ((match = emojiExp.exec(emojiText)) !== null) {
+     results.push(emojiText.substring(idx, match.index))
+     results.push({type: 'native-emoji', children: [emojiIndex[match[0]]]})
+     idx = match.index + match[0].length
+   }
+   results.push(emojiText.substring(idx, emojiText.length))
+   return results.filter(Boolean)
+ }
 
-Blank
-  = ws:(WhiteSpace / LineTerminatorSequence) {
-      return ws;
-    }
+LinkSpecialChar
+ = EscapeMarker / StrikeMarker / BoldMarker / ItalicMarker
+
+LinkChar
+ = !LinkSpecialChar char:NonBlank { return char }
+
+Link
+ = proto:("http"i "s"i? ":")? url:(LinkChar+) & {
```

Comment:
This now uses a regex lookahead for blocks of text that could be a url. In the future, we could probably make this faster by moving some of the domain parsing logic from the regex into this grammar rule.